### ISSUE TITLE
Add pagination to the API?

### ISSUE THREAD

**Issue by peterdesmet** [ices-tools-dev/esas#120]
https://esas.ices.dk/api/getPositionRecords loads and then crashes in my browser, I assume because it is trying to return too many records. It still works in R (not trying to load in browser):

```
positions <- fromJSON("https://esas.ices.dk/api/getPositionRecords")
```

Do you think we should introduce pagination for the results? See e.g. section "paging" of the GBIF API, which has parameters `offset` and optionally `limit`:

- offset: 0, limit: 100: first page with 100 results
- offset: 0, limit: 20: first page with 20 results
- offset: 20, limit 20: second page with 20 results (results 21-40)

If we don't introduce pagination, it would probably be good to remove the examples that don't filter results on the API documentation page at https://esas.ices.dk/webservices

**Comment by peterdesmet** [comment]
Remove unfiltered examples from API doc page

**Comment by neil-ices-dk** [comment]
> unfiltered examples from API doc page

@cmspinto to remove the unfiltered examples; wait with pagination for the time being

**Comment by peterdesmet** [comment]
Now that there's lots of data, the API response is slow (see below). I'd like to re-suggest pagination. 😄
 
 - campaigns: 1.040714 mins
 - samples: 40.10294 secs
 - positions: 8.440281 mins
 - observations: 30.07193 mins

**Comment by cmspinto** [comment]
Pagination is implemented in the WebAPI.

https://esas.ices.dk/webservices

![image](https://user-images.githubusercontent.com/8705477/198031679-797d9eab-da78-4fd7-9c2e-a783b6725625.png)


@peterdesmet can you check if this is what you expected?

**Comment by peterdesmet** [comment]
Thanks @cmspinto. I assume the records are always ordered the same way?

I think the functionality is there, but it took me a bit to understand that the notation is`offset/limit/`, especially since `/` is often used for range `start/end`. Another disadvantage is that one cannot query without these parameters anymore: https://esas.ices.dk/api/getCampaignRecords?country=dk

I would therefore make these query parameters `?offset=0&limit=1000`. That has the advantage:

- More explicit
- Optional (so users can try API without it)

We can even simplify further (e.g. what if the user
specifies a `offset`, but no `limit`) and not allow a custom`limit`, but hardcode it at 1000. So:

- https://esas.ices.dk/api/getCampaignRecords?country=dk: all records for DK
- https://esas.ices.dk/api/getCampaignRecords?country=dk&offset=0: first 1000 records for DK
- https://esas.ices.dk/api/getCampaignRecords?country=dk&offset=1000: records 1001-2000 for DK

Typically, an API with pagination returns the offset, total count, and results, not only the results. See: https://api.gbif.org/v1/occurrence/search?taxonKey=1

```
{"offset":0,"limit":20,"endOfRecords":false,"count":1794128283,"results":[]}
```

**Comment by cmspinto** [comment]
@peterdesmet can move the pagination to optional. From the url path to the querystring.

I'm a bit more reluctant to make it part of the results, being the reason that it makes only a bit more difficult for users to use the data.

Right now a user can just download and use the data (event in excel).

If we put a summary on the top,  and data as part of the results then it makes it even a bit more technical.

Just to note that I think that the solution you are suggesting is the best one, the most clear and the correct one, I just wonder if that won't bring some trouble to part of the users (that use excel or other software)?

**Comment by cmspinto** [comment]
@peterdesmet , it's done. Takes a bit longer to query but still under the acceptable (I think).

![image](https://user-images.githubusercontent.com/8705477/198255017-d435a741-7d7f-4857-9153-4ecb9f5a4345.png)

or 

![image](https://user-images.githubusercontent.com/8705477/198255347-b906abcd-6b9f-4af9-b306-88d628951d42.png)


I will wait to publish the API and the website, because @nicolasvanermen is submitting files and this could affect the submissions.

**Comment by peterdesmet** [comment]
> @peterdesmet , it's done.
Nice, I think the result set is a more future-proof approach, as it is less likely to break things if changes are made to the API.

> I will wait to publish the API and the website, because @nicolasvanermen is submitting files and this could affect the submissions.
Ok, let me know! Looking forward to test this.

> The offset and the limit are not mandatory, and follow your suggestion, if the offset is declared but not limit then it takes 1000 for default.
Great!

**Comment by cmspinto** [comment]
@peterdesmet go ahead, fortunately it's a bit faster in the server than it was in my computer

**Comment by peterdesmet** [comment]
I'm using a different internet connection, so not a robust benchmark, but working with chunks has improved speed: :+1:

campaigns: 1.040714 mins -> 7.000045 secs
samples: 40.10294 secs -> 13.03842 secs
positions: 8.440281 mins -> 1.633067 mins

For observations however, I'm still noticing very slow (to slower) queries. Might this have something to do with the fact that a  `recordsCount` needs to be created for every query?

**Comment by cmspinto** [comment]
Yes, that is exactly it. 

It needs to replicate the query and check how many records are in the result.

Do you know what is lazy loading ([https://learn.microsoft.com/en-us/ef/core/querying/related-data/lazy](https://learn.microsoft.com/en-us/ef/core/querying/related-data/lazy))? 
Lazy loading is not  really working anymore because it needs to count the records before if returns the results!

**Comment by peterdesmet** [comment]
So `recordsCount` prevents lazy loading. What about `endOfRecords`?

**Comment by cmspinto** [comment]
It's the same, it know if it reach the end of the records by doing the count. 
If the user ask for 20 records, it needs to do the count to be able to know if it is returning all the records or if there is any more records after the 20th one.

**Comment by peterdesmet** [comment]
In that case it might be better to drop the calculation of those two fields, so we can enjoy the speed of lazy loading again?

**Comment by cmspinto** [comment]
@peterdesmet , remove those two fields is an option, but it's more accurate and it's a bit more clear to the end user right?

I'm fine with both options, but I can see your point.

**Comment by peterdesmet** [comment]
Yes, those fields are useful, but I’d rather have a fast API 😊. They will still see the returned offset and limit, which indicates pagination.

I would thus remove `recordCount` and `endOfRecords` to speed things up. My script/function will be code around it (basically try offset until no more records are returned).

**Comment by peterdesmet** [comment]
@cmspinto you mentioned in https://github.com/ices-tools-dev/esas/issues/135#issuecomment-1296861883 that lazy loading is back.

I'm however still getting (as) slow responses from e.g. https://esas.ices.dk/api/getObservationRecords?offset=1000&limit=1000 Any idea what is going on?

**Comment by cmspinto** [comment]
I agree Peter, when no filter is put in getObservationRecords then the results always take around 19s to load. 

A possible alternative is to pull the data by campaignID: 
[[campaignID](https://esas.ices.dk/api/getObservationRecords?offset=0&limit=1000&campaignID=775)](https://esas.ices.dk/api/getObservationRecords?offset=0&limit=1000&campaignID=775)
Then it takes about 178ms.

![image](https://user-images.githubusercontent.com/8705477/199464299-b380763b-f903-4a3d-aeda-b18413ad105e.png)

I would guess it is the internals of the object. 

It is still far from the minutes it took when it had to do the count of the records.

**Comment by cmspinto** [comment]
I was looking and the Observations is a union of selects, this is because Species can we reported using Worms or ESAS. 
I might be able to optimize a little bit this download.

**Comment by cmspinto** [comment]
The time is down to 4.7s

![image](https://user-images.githubusercontent.com/8705477/199673388-4b39a590-3a08-4fde-a6b7-814740406c12.png)


What is keeping it in 4.7s is the function to pull the vernacular name. This query can take less than a second if it does not need to search for an English vernacular name using a function.

@peterdesmet  is 4.7s fine?

**Comment by cmspinto** [comment]
Can we close this issue?
